.github/workflows/build.yml

```
@@ -0,0 +1,64 @@
+name: Build struct2tensor wheels
+
+on:
+  pull_request:
+  workflow_dispatch:
+  release:
+    types: [published]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.9", "3.10", "3.11"]
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Build package
+        id: build-package
+        uses: ./.github/reusable-build
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: List and check wheels
+        shell: bash
+        run: |
+          # Pin transitive dependency on pkginfo until https://github.com/pypa/twine/issues/1070 is
+          # fixed
+          pip install twine pkginfo>=1.10.0
+          ${{ matrix.ls || 'ls -lh' }} dist/
+          twine check dist/*
+
+  upload_to_pypi:
+    name: Upload to PyPI
+    runs-on: ubuntu-latest
+    if: (github.event_name == 'release' && startsWith(github.ref, 'refs/tags')) || (github.event_name == 'workflow_dispatch')
+    needs: [build]
+    environment:
+      name: pypi
+      url: https://pypi.org/p/struct2tensor/
+    permissions:
+      id-token: write
+    steps:
+      - name: Retrieve wheels
+        uses: actions/download-artifact@v4
+        with:
+          merge-multiple: true
+          path: wheels
+
+      - name: List the build artifacts
+        run: |
+          ls -lAs wheels/
+
```

```suggestion
      - name: List the build artifacts
        run: |
          ls -lAs wheels/

     - name: Generate artifact attestations for wheels
       uses: actions/attest-build-provenance@v1.4.3
       with:
         subject-path: "wheels/*"
```

It might be good to use the GHA Attestations feature to establish SLSA provenance of the built wheels ([GitHub's blog post](https://github.blog/changelog/2024-06-25-artifact-attestations-is-generally-available/));, they can then be verified using [the `gh attestation verify` command](https://cli.github.com/manual/gh_attestation_verify) and will be stored permanently in https://github.com/google/struct2tensor/atttestations/.

---

It seems like this is something that can only help our users, and I'm happy to add this here. Thank you for the links - they provided plenty of context.